lhotse/dataset/sampling.py

```diff
@@ -0,0 +1,322 @@
+import logging
+import math
+import random
+import warnings
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, List, Optional, Tuple
+
+import torch
+
+from lhotse import CutSet
+
+
+class CutSampler(ABC):
+    """
+    CutSampler is responsible for collecting batches of cuts, given specified criteria.
+    It implements correct handling of multiprocessing in DataLoader,
+    as well as distributed training, so that the cuts are not duplicated across workers.
+
+    Sampling in a CutSampler is intended to be very quick - it only uses the metadata in
+    ``CutSet`` manifest to select the cuts, and is not intended to perform any I/O.
+
+    CutSampler works similarly to PyTorch's DistributedSampler - when :attr:`shuffle=True`,
+    you should call ``sampler.set_epoch(epoch)`` at each new epoch to have a different
+    ordering of returned elements.
+
+    Example usage::
+
+        >>> dataset = K2SpeechRecognitionIterableDataset(
+        ...     cuts,
+        ...     sampler=SingleCutSampler(cuts, world_size=4, local_rank=0)
+        ... )
+        >>> loader = DataLoader(dataset, batch_size=None)
+        >>> for epoch in range(start_epoch, n_epochs):
+        ...     dataset.sampler.set_epoch(epoch)
```

Comment:
this set_epoch stuff, I suppose we'd need to change the snowfall scripts..
although I believe I've seen a paper saying that using a deterministic order can work better.

Reply:
Interesting, could you share the reference?

Reply:
Sorry couldn't find it.